packages/sdk/src/client.ts

```diff
@@ -677,6 +723,211 @@ export class IndexerClient {
+		while (!item.done) {
+			yield item.value
+			item = await combined.next()
+		}
+		return
+	}
+
+	/**
+	 * Create a Stream of status updates
+	 * @param hash - Can be commitment, hyperbridge tx hash, source tx hash, destination tx hash, or timeout tx hash
+	 * @returns AsyncGenerator that emits status updates until a terminal state is reached
+	 */
+	private async *getRequestStatusStreamInternal(hash: string): AsyncGenerator<RequestStatusWithMetadata, void> {
+		const self = this
+		let request: GetRequestWithStatus | undefined
+		while (!request) {
+			await sleep(self.config.pollInterval)
+			request = await self.queryGetRequest(hash)
+		}
+
+		let status =
+			request.source === self.config.hyperbridge.stateMachineId
+				? RequestStatus.HYPERBRIDGE_DELIVERED
+				: RequestStatus.SOURCE
+		const latestMetadata = request.statuses[request.statuses.length - 1]
+		// start with the latest status
+		status = maxBy(
+			[status, latestMetadata.status as RequestStatus],
+			(item) => REQUEST_STATUS_WEIGHTS[item as RequestStatus],
+		)!
+
+		while (true) {
+			switch (status) {
+				// request has been dispatched from source chain
+				case RequestStatus.SOURCE: {
+					let sourceUpdate: StateMachineUpdate | undefined
+					while (!sourceUpdate) {
+						await sleep(self.config.pollInterval)
+						sourceUpdate = await self.queryStateMachineUpdateByHeight({
+							statemachineId: request.source,
+							height: request.statuses[0].metadata.blockNumber,
+							chain: self.config.hyperbridge.stateMachineId,
+						})
+					}
+
+					yield {
+						status: RequestStatus.SOURCE_FINALIZED,
+						metadata: {
+							blockHash: sourceUpdate.blockHash,
+							blockNumber: sourceUpdate.height,
+							transactionHash: sourceUpdate.transactionHash,
+						},
+					}
+					status = RequestStatus.SOURCE_FINALIZED
+					break
+				}
+
+				// finality proofs for request has been verified on Hyperbridge
+				case RequestStatus.SOURCE_FINALIZED: {
+					// wait for the request to be delivered on Hyperbridge
+					while (!request || request.statuses.length < 2) {
+						await sleep(self.config.pollInterval)
+						request = await self.queryGetRequest(hash)
+					}
+
+					status =
+						request.dest === self.config.hyperbridge.stateMachineId
+							? RequestStatus.DESTINATION
+							: RequestStatus.HYPERBRIDGE_DELIVERED
+
+					yield {
+						status,
+						metadata: {
+							blockHash: request.statuses[1].metadata.blockHash,
+							blockNumber: request.statuses[1].metadata.blockNumber,
+							transactionHash: request.statuses[1].metadata.transactionHash,
+						},
+					}
+					break
+				}
+
+				// the request has been verified and aggregated on Hyperbridge
+				case RequestStatus.HYPERBRIDGE_DELIVERED: {
+					// Get the latest state machine update for hyperbridge on the destination chain
+					let hyperbridgeFinalized: StateMachineUpdate | undefined
+					let index = request.source === self.config.hyperbridge.stateMachineId ? 0 : 1
```

Comment:
i don't think this makes sense here. If hyperbridge was the source, then this request has already completed at `HYPERBRIDGE_DELIVERED`. You can simply terminate the generator. Otherwise we want the height at which it was processed on hyperbridge which should be in the second status, index 1